src/crimson/osd/osd.cc

```diff
@@ -775,12 +775,18 @@ OSD::ms_dispatch(crimson::net::ConnectionRef conn, MessageRef m)
           srand((unsigned)time(NULL));
           insert_iter->second = rand()% seastar::smp::count;
         }
-        return conn.get_foreign().then([this, m = std::move(m),
-          core = insert_iter->second](auto f_conn) {
-          return shard_dispatchers.invoke_on(core,
```

Comment:
`conn.get_foreign()` cannot guarantee the ordering here, so it may not be caused by `invoke_on()`.

If `invoke_on()` or `smp::submit_to()` cannot guarantee fifo in the submitted shard, it would be a larger issue. We probably need to scan all the places that is using them.

Reply:
I just took a quick look at the `smp::submit_to()` implementation, looks it is operating a `pending_fifo` queue explicitly, which strongly implies that it enforces fifo.

So I don't think it worths protecting `invoke_on()` or `smp::submit_to()` with additional cross-core unlock mechanisms.

@athanatos @liu-chunmei  what do you think?

Reply:
Seems reasonable to me, leave a comment though.

Reply:
seems I only need to guarantee conn.get_foreign() ordering, will modify it.

Reply:
@cyx1231st @athanatos if I only add mutex to guarantee conn.get_foreign() ordering, still has out-of order problem. see below patch, doesn't work at all.
+        return op_mutex.lock().then([=, this] {
+          return conn.get_foreign().then([this, m = std::move(m),
+            core = insert_iter->second](auto f_conn) {
+            op_mutex.unlock();
+            return shard_dispatchers.invoke_on(core,
+              [f_conn = std::move(f_conn), m = std::move(m)]
+              (auto &local_dispatcher) mutable ->seastar::future<>{
+              return local_dispatcher.ms_dispatch(std::move(f_conn), m);
+            });

Reply:
@athanatos @cyx1231st if I only add mutex to guarantee conn.get_foreign() order, still has out-of-ordering problem. see below PR, still has out-of-ordering issue.
---
+        return op_mutex.lock().then([=, this] {
+          return conn.get_foreign().then([this, m = std::move(m),
+            core = insert_iter->second](auto f_conn) {
+            op_mutex.unlock();
+            return shard_dispatchers.invoke_on(core,
+              [f_conn = std::move(f_conn), m = std::move(m)]
+              (auto &local_dispatcher) mutable ->seastar::future<>{
+              return local_dispatcher.ms_dispatch(std::move(f_conn), m);
+            });
---

Reply:
I got this ordering with only lock conn.get_foreign, the same problem as without any lock here. 
primary core ms_dispatch sequence
---osd::ms_dispatch version = 0'0---reqid = 11735
---osd::ms_dispatch version = 0'0---reqid = 11736

shardDispatcher: handle_rep_op seqeunce
---osd::handle_rep_op version = 19'98---reqid = 11736---shard id = 1
---osd::handle_rep_op version = 19'97---reqid = 11735---shard id = 1

Reply:
print location: 
```
   in osd::ms_dispatch: 
     switch (m->get_type()) {
+      case MSG_OSD_REPOP:
+std::cout<<"---osd::ms_dispatch version = "<<boost::static_pointer_cast<MOSDRepOp>(m)->version<<"---reqid = "<<boost::static_pointer_cast<MOSDRepOp>(m)->reqid.tid<<std::endl;
+      [[fallthrough]];

OSD::ShardDispatcher::handle_rep_op(
   m->finish_decode();
+std::cout<<"---osd::handle_rep_op version = "<<m->version<<"---reqid = "<<m->reqid.tid<<"---shard id = "<<seastar::this_shard_id()<<std::endl;
   std::ignore = pg_shard_manager.start_pg_operation<RepRequest>(
     std::move(conn),
     std::move(m));
```

Reply:
There is a semaphore in `invoke_on()`/`submit_to()`, which cannot grarantee the FIFO order at `wait()`, quoted "Waits are serviced in FIFO order, though if several are awakened at once, they may be reordered by the scheduler."

So we shouldn't expect the target core gets called in the same order. I have also observed the same issue during multi-core messenger testing, which needs to be addressed separately.

Another impact is that we need to chain the future returned by `Connection::send()` wherever there is an ordering requirement.